modules/biobank/php/pooldao.class.inc

```diff
@@ -0,0 +1,307 @@
+                  FROM biobank_pool';
+        $query .= $conditions ? ' WHERE '.$this->db->implodeWithKeys(' AND ', $conditions) : '';
+
+        $result = $this->db->pselect($query, array());
+        $pools  = array();
+        if (!empty($result)) {
+            foreach ($result as $row) {
+                $id         = intval($row['PoolID']);
+                $pool       = $this->getInstanceFromId($id);
+                $pools[$id] = $pool;
+            }
+        }
+
+        return $pools;
+    }
+
+    /**
+     * Instantiates an ArrayIterator class that is composed of all the Pool
+     * Objects
+     *
+     * @return traversable Iterator of Pool Objects
+     */
+    protected function getAllInstances() : \Traversable
+    {
+        return new \ArrayIterator($this->selectInstances());
+    }
+
+    /**
+     * This function recieves a Pool Object, converts it into an SQL format
+     * and inserts it into all relevant tables in the database. Certain
+     * information is then cascaded to the children of the Pool.
+     *
+     * @param Pool $pool The Pool to be inserted into the database.
+     *
+     * @return void
+     */
+    public function saveInstance(Pool $pool) : void
+    {
+        //Convert Pool to SQL format.
+        $poolArray = getArrayFromInstance();
+        $data = $this->getSQLFromArray($poolArray);
+
+        // TODO: Introduce a SQL Transaction here with a try catch.
+        // Insert or Update the biobank_pool table with respective data.
+        $this->db->insertOnDuplicateUpdate(
+            'biobank_pool',
+            $data['biobank_pool']
+        );
+
+        $poolId = $pool->getId();
+        if (isset($poolId)) {
+            // If update:
+            // XXX: Logic will likely eventually go here.
+        } else {
+            // If insert:
+            $poolId = $this->db->getLastInsertId();
+            // set the respective data array index to the value of the new
+            // pool ID
+            $specimenIds = $pool->getSpecimenIds();
+            foreach ($specimenIds as $id) {
+                $data['biobank_specimen_pool_rel'][$id]['PoolID'] = $poolId;
+            }
+        }
+
+        // insert or update biobank_specimen_pool_rel with relevant data.
+        foreach ($data['biobank_specimen_pool_rel'] as $insert) {
+            $this->db->insertOnDuplicateUpdate(
+                'biobank_specimen_pool_rel',
+                $insert
+            );
+        }
+
+        // XXX: decide if quantity of all pooled specimens should go to 0.
+        // XXX: logic for adding preparations to pools will likely go here as well.
+    }
+
+    /**
+     * This method takes a Pool Object and prepares the data to be inserted into
+     * the database by converting it to a data array.
+     *
+     * @param Pool $pool Pool Object to be converted
+     *
+     * @return array $data Array containing the data to be inserted
+     */
+    private function getSQLFromArray(array $pool) : array
```

Comment:
```suggestion
    private function getSQLFromArray(Pool $pool) : array
```

Reply:
This actually takes an array, and not a pool object! I'll change the function documentation.